[jest] Add watch plugins (#5563)
expo/expo: packages/jest-expo/README.md

You can use `jest-expo` to test any Expo supported platform. For legacy purposes `jest-expo` runs your tests in the standard React Native environment (iOS).
The recommended way to test your project is with `jest-expo/universal` which runs your tests with every Expo supported platform. Currently this includes iOS, Android, web, and Node (which is used for testing SSR compliance).

> Pressing **X** will open a platform selection dialog which you can use to debug single platforms. You can also create a custom `jest.config` and combine the single platforms with `jest-expo/ios`, `jest-expo/android`, `jest-expo/web`, and `jest-expo/node`.

Reviewer: ```suggestion
Pressing **X** will open a platform-selection dialog that you can use to test individual platforms. You can also create a custom Jest config and combine the individual platforms with `jest-expo/ios`, `jest-expo/android`, `jest-expo/web`, and `jest-expo/node`.
```

- web: `-test.web.*`
- Node: `-test.node.*`, `-test.web.*`

### ⚙️ Configuring your preset

When building a custom preset you may want to use some of features provided by this preset. You can access these features through the `jest-expo/config` directory.

#### `getWatchPlugins(jestConfig)`

> When given an existing Jest config this will return the `watchPlugins` used in `jest-expo`. This reads the `projects` field to return watchPlugins for single-project and multi-project configs.

Reviewer: ```suggestion
When given an existing Jest config this will return the `watchPlugins` used in `jest-expo`. This reads the `projects` field to determine which watchPlugins to return for single-project and multi-project configs.
```

> Currently this returns type-ahead for all projects

Reviewer: ```suggestion
Currently this returns type-ahead plugins for all projects:
```

- `jest-watch-typeahead/filename`
- `jest-watch-typeahead/testname`

> And a custom platform selection dialog for multi-projects

Reviewer: ```suggestion
And a custom platform selection dialog for universal multi-projects:
```

- `jest-watch-select-projects`

#### `withWatchPlugins(jestConfig)`

Given a Jest config, this will ensure any existing `watchPlugins` are safely merged with `getWatchPlugins(jestConfig)`.

#### `getWebPreset()`

Alternative to `jest-expo/web`. This runs in a JSDOM environment for testing **Expo web**.

#### `getIOSPreset()`

Alternative to `jest-expo/ios`. Runs in a mock native environment.

#### `getAndroidPreset()`

Alternative to `jest-expo/android`. Also runs in a mock native environment.

#### `getNodePreset()`

> Alternative to `jest-expo/node`. This runs in a node environment for testing **SSR**.

Reviewer: ```suggestion
Alternative to `jest-expo/node`. This runs in a Node environment for testing **SSR**.
```
